### ISSUE TITLE
groupby_dynamic: rows are dropped; boundaries and offsets are incorrect

### ISSUE THREAD

**Issue by cbilot** [pola-rs/polars#3776]
polars 0.13.49
python 3.10.4
Linux Mint 20.3

#### Describe your bug.
When using `groupby_dynamic`, rows of the original dataset are being dropped, and the offsets and the boundaries seem incorrect.

#### What are the steps to reproduce the behavior?
If we start with this data...
```python3
from datetime import datetime

import polars as pl

df = pl.DataFrame(
    {"date": [datetime(2020, 12, 1), datetime(2021, 2, 1), datetime(2021, 5, 1)]}
)
df
```
```
shape: (3, 1)
┌─────────────────────┐
│ date                │
│ ---                 │
│ datetime[μs]        │
╞═════════════════════╡
│ 2020-12-01 00:00:00 │
├╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌┤
│ 2021-02-01 00:00:00 │
├╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌┤
│ 2021-05-01 00:00:00 │
└─────────────────────┘
```
## "1d"
Performing a `groupby_dynamic` with `every="1d"` produces the expected result:
```python
df.groupby_dynamic("date", "1d", include_boundaries=True).agg(
    pl.col("date").alias("date_list")
)
```
```
shape: (3, 4)
┌─────────────────────┬─────────────────────┬─────────────────────┬───────────────────────┐
│ _lower_boundary     ┆ _upper_boundary     ┆ date                ┆ date_list             │
│ ---                 ┆ ---                 ┆ ---                 ┆ ---                   │
│ datetime[μs]        ┆ datetime[μs]        ┆ datetime[μs]        ┆ list[datetime[μs]]    │
╞═════════════════════╪═════════════════════╪═════════════════════╪═══════════════════════╡
│ 2020-11-30 00:00:00 ┆ 2020-12-01 00:00:00 ┆ 2020-12-01 00:00:00 ┆ [2020-12-01 00:00:00] │
├╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌┤
│ 2021-01-31 00:00:00 ┆ 2021-02-01 00:00:00 ┆ 2021-02-01 00:00:00 ┆ [2021-02-01 00:00:00] │
├╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌┤
│ 2021-04-30 00:00:00 ┆ 2021-05-01 00:00:00 ┆ 2021-05-01 00:00:00 ┆ [2021-05-01 00:00:00] │
└─────────────────────┴─────────────────────┴─────────────────────┴───────────────────────┘
```
Per the documentation for `offset`:
> offset of the window if None and period is None it will be equal to negative every

The offset is automatically set to negative one day, and the boundaries are correct.
## "2d"
However, if we change the `every` keyword to two days...
```python3
df.groupby_dynamic("date", "2d", include_boundaries=True).agg(
    pl.col("date").alias("date_list")
)
```
```
shape: (2, 4)
┌─────────────────────┬─────────────────────┬─────────────────────┬───────────────────────┐
│ _lower_boundary     ┆ _upper_boundary     ┆ date                ┆ date_list             │
│ ---                 ┆ ---                 ┆ ---                 ┆ ---                   │
│ datetime[μs]        ┆ datetime[μs]        ┆ datetime[μs]        ┆ list[datetime[μs]]    │
╞═════════════════════╪═════════════════════╪═════════════════════╪═══════════════════════╡
│ 2021-01-30 00:00:00 ┆ 2021-02-01 00:00:00 ┆ 2021-01-31 00:00:00 ┆ [2021-02-01 00:00:00] │
├╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌┤
│ 2021-04-30 00:00:00 ┆ 2021-05-02 00:00:00 ┆ 2021-05-01 00:00:00 ┆ [2021-05-01 00:00:00] │
└─────────────────────┴─────────────────────┴─────────────────────┴───────────────────────┘
```
The first row of our dataset has been dropped.  And the `date` value in the first row of our results is not a value in our original dataset.  (Not sure where this value is coming from.)
## "3d"
```python3
df.groupby_dynamic("date", "3d", include_boundaries=True).agg(
    pl.col("date").alias("date_list")
)
```
```
shape: (2, 4)
┌─────────────────────┬─────────────────────┬─────────────────────┬───────────────────────┐
│ _lower_boundary     ┆ _upper_boundary     ┆ date                ┆ date_list             │
│ ---                 ┆ ---                 ┆ ---                 ┆ ---                   │
│ datetime[μs]        ┆ datetime[μs]        ┆ datetime[μs]        ┆ list[datetime[μs]]    │
╞═════════════════════╪═════════════════════╪═════════════════════╪═══════════════════════╡
│ 2021-01-30 00:00:00 ┆ 2021-02-02 00:00:00 ┆ 2021-01-30 00:00:00 ┆ [2021-02-01 00:00:00] │
├╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌┤
│ 2021-04-30 00:00:00 ┆ 2021-05-03 00:00:00 ┆ 2021-04-30 00:00:00 ┆ [2021-05-01 00:00:00] │
└─────────────────────┴─────────────────────┴─────────────────────┴───────────────────────┘
```
The first row is dropped.  And neither of the `date` values are from the original dataset.

**Edit** I think I've figured out how `groupby_dynamic` calculates intervals in months.

**Comment by cbilot** [comment]
Okay, after playing around with the every/period/offset values in months, I may have stumbled onto the answer.  

When using months, does the algorithm always start calculating intervals at the beginning of the year of the time index of the first record, and then calculate intervals from that starting value?

For example, if we have this data:
```python3
from datetime import datetime

import polars as pl

df = pl.DataFrame(
    {"date": [
        datetime(2020, 2, 11),
        datetime(2021, 2, 11),
        datetime(2022, 2, 11),
    ]}
)
df
```
```
shape: (3, 1)
┌─────────────────────┐
│ date                │
│ ---                 │
│ datetime[μs]        │
╞═════════════════════╡
│ 2020-02-11 00:00:00 │
├╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌┤
│ 2021-02-11 00:00:00 │
├╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌┤
│ 2022-02-11 00:00:00 │
└─────────────────────┘
```
Then, for `every="5mo"`, we start at `2020-01-01` and then proceed with 5 month intervals after that?
```
df.groupby_dynamic("date", "5mo", include_boundaries=True).agg(
    pl.col("date").alias("date_list")
)
```
```
shape: (3, 4)
┌─────────────────────┬─────────────────────┬─────────────────────┬───────────────────────┐
│ _lower_boundary     ┆ _upper_boundary     ┆ date                ┆ date_list             │
│ ---                 ┆ ---                 ┆ ---                 ┆ ---                   │
│ datetime[μs]        ┆ datetime[μs]        ┆ datetime[μs]        ┆ list[datetime[μs]]    │
╞═════════════════════╪═════════════════════╪═════════════════════╪═══════════════════════╡
│ 2020-01-01 00:00:00 ┆ 2020-06-01 00:00:00 ┆ 2020-01-01 00:00:00 ┆ [2020-02-11 00:00:00] │
├╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌┤
│ 2020-11-01 00:00:00 ┆ 2021-04-01 00:00:00 ┆ 2020-11-01 00:00:00 ┆ [2021-02-11 00:00:00] │
├╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌┤
│ 2022-02-01 00:00:00 ┆ 2022-07-01 00:00:00 ┆ 2022-02-01 00:00:00 ┆ [2022-02-11 00:00:00] │
└─────────────────────┴─────────────────────┴─────────────────────┴───────────────────────┘
```
If that's true, then the answer to the Stack Overflow question:
>Is every='3mo' equal to pandas resample('Q') ?

is actually 'yes'.  (The other part of the question is just using `when/then/otherwise` and `polars.datetime` to derive the quarter-ending date from the `date` field, so that using `include_boundaries=True` isn't necessary.)

In retrospect, I s'pose when dealing with months, there really is no other easy way to count in months, especially since there are a different number of days in each month, so that subtracting "one month" from `2020-03-30` doesn't make much sense because there is no `2020-02-30`.

Still, I can't figure out the issue with "1d", "2d", "3d".  Is that a bug?  (Sorry for making a disaster of this, but it did take me quite some time to figure out the "months" thing.)

**Comment by ritchie46** [comment]
I am looking into it. I might edit this answer as I go. ;)

> The first row of our dataset has been dropped. And the date value in the first row of our results is not a value in our original dataset. (Not sure where this value is coming from.)

`truncate` is set to `True` which truncates the `date` column to the lower bound.

**Comment by ritchie46** [comment]
I have got a fix for the dropped rows for days in #3779. I am not really sure about what to do in months yet.

> In retrospect, I s'pose when dealing with months, there really is no other easy way to count in months, especially since there are a different number of days in each month, so that subtracting "one month" from 2020-03-30 doesn't make much sense because there is no 2020-02-30.

Yes, for months we do something totally different. We have to deal with different month lengths and leap years, so we only give month boundaries start from start to end of month. Still it might make sense to start at the data location with the months instead of start of the year?

Which would make more sense?

**Comment by ritchie46** [comment]
Fixed by #3779
